### PR TITLE
Use dtype in pandas to evaluate type where possible

### PR DESCRIPTION
**Pull request by jcampbell**
Hold this for after @RoyalTS PRs merge; another in the same spirit.

### CODE REVIEW

File: great_expectations/dataset/pandas_dataset.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 import inspect
 import json
 import re
-from datetime import datetime
+from datetime import datetime # , timedelta # Add for case of testing timedelta types
```
Comment (abegong) [inline great_expectations/dataset/pandas_dataset.py]: ?

File: great_expectations/dataset/pandas_dataset.py [hunk 2]
```diff
@@ -520,12 +520,24 @@ def expect_column_values_to_be_of_type(self, column, type_,
             "float": [float, np.float_],
             "double": [float, np.longdouble],
             "bytes": [bytes, np.bytes_],
-            "string": [string_types, np.string_]
+            "string": [np.string_] + list (string_types)
```
Comment (abegong) [inline great_expectations/dataset/pandas_dataset.py:447]: Yikes. How did this get past us before?

**Review by abegong — COMMENTED** [review]
Does this need tests?

**Comment by RoyalTS** [comment]
I'm not sure what to think of this. On the one hand, it makes a ton of sense to be able to test for dtypes. But it also makes sense to be able to test for underlying types (e.g. "are all the things in that `object` column strings?). If you're testing dtypes, I would suggest to test using the pandas board methods for that sort of thing, e.g. https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.api.types.is_string_dtype.html

File: great_expectations/dataset/pandas_dataset.py (hunk 2 above)
Comment (jcampbell) [reply great_expectations/dataset/pandas_dataset.py:447]: Before, the code always called `tuple()` on the list as part of the lambda, and use `isinstance`. Now, in the event that the whole column is of a specific dtype we bypass that check entirely and rely on `in` which requires the list and doesn't re-call tuple.
image

File: great_expectations/dataset/pandas_dataset.py (hunk 1 above)
Comment (jcampbell) [reply great_expectations/dataset/pandas_dataset.py]: Added as a note about the placeholder for potentially adding additional types accepted by this expectation, but thought of it as out of scope for the current PR (which is optimization focused)

**Comment by jcampbell** [comment]
Great point on the change to using the types api, will do.

I also agree that it's important to be able to test for underlying types...but I think this PR still does that by first checking whether the column dtype is 'object' and, if so, checking row-wise. Does that address that issue?

**Comment by RoyalTS** [comment]
I've become very wary of mixing types and dtypes. Would keep them wholly separate if it were up to me.

**Comment by jcampbell** [comment]
Fair enough...I was just scanning through places where we could short circuit looping over everything in the column after your comments. I'd say the current implementation, especially by relying on per-implementation type-lists is already too unwieldy (and we've discussed that in issues in the past). I'll just close this PR in favor of reviewing type expectations more comprehensively at some point in the future.

**Pull request merged**